### PR TITLE
parser_v2

### PR DESCRIPTION
**Pull request by jxt1009**
A different approach at the structure of the new nodes, appears to work a lot better.
Also solved issues with reassignment not working. Everything appears to pass as expected now
The output of the json data looks great too.

### CODE REVIEW

File: Parser/Tree_Nodes/Parse_Nodes/for_node.c [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+for_node *create_for_node(asmt *initalizer, i_expr *conditional, r_asmt *incrementer, struct b_stmt_list *body) {
+    for_node *retVal = NULL;
+    retVal = calloc(1, sizeof(for_node));
+    retVal->initialize = initalizer;
+    retVal->conditional = conditional;
+    retVal->incrementer = incrementer;
+    retVal->body = body;
+    return retVal;
+}
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/for_node.c]: What is going on here? Shouldn't the constructor take in the token stream, index and next pointer like every other constructor? If you need to deviate from the pattern, create a helper that the constructor calls, like in i_expr

File: Parser/Tree_Nodes/Parse_Nodes/for_node.c [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+GString *for_node_to_json(for_node *forNode) {
+    GString *retval = g_string_new(NULL);
+
+    if (retval != NULL) {
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/for_node.c]: checking forNode would be more useful

File: Parser/Tree_Nodes/Parse_Nodes/for_node.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+} for_node;
+
+/**
+ * @brief Creates an assignment from the token_stream starting at the given index and updates next to be the following
+ * token index.
+ * @param token_stream The stream of tokens for the entire program.
+ * @param index The index in the token stream to start building an assignment from.
+ * @param next[OUT] A pointer to the index of the next token not part of the assignment.
+ * @return A for loop held in a node
+ */
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/for_node.h]: Copy Pasta?

File: Parser/Tree_Nodes/Parse_Nodes/if_node.c [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+
+if_node *create_if_node(i_expr *conditional, struct b_stmt_list *true_branch, struct b_stmt_list *false_branch) {
+    if_node *retVal = NULL;
+    retVal = calloc(1, sizeof(if_node));
+    retVal->expression = conditional;
+    retVal->b_true = true_branch;
+    retVal->b_false = false_branch;
+    return retVal;
+}
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/if_node.c]: This also does not follow the contsturctor pattern

File: Parser/Tree_Nodes/Parse_Nodes/while_node.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+while_node *create_while_node(i_expr *conditional, b_stmt_list *body) {
+    while_node *retVal = NULL;
+
+    // Make new While node
+    retVal = calloc(1, sizeof(while_node));
+    retVal->body = body;
+    retVal->conditional = conditional;
+    //Return the value
+    return retVal;
+}
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/while_node.c]: This also does not follow the constructor pattern

File: Parser/Tree_Nodes/Parse_Nodes/while_node.h [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+typedef struct while_node {
+    i_expr *conditional;
+    struct b_stmt_list *body;
+} while_node;
+
+/**
+ * @param value[IN] A token of the condition to continue the loop
+ * @param sign[IN] A statement list to run while the condition is true
+ * @return The new while_node
+ */
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/Parse_Nodes/while_node.h]: This doesn't match anything

File: Parser/Tree_Nodes/b_stmt.c [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+            new_statement->expression = create_expr(token_stream, index, next);
+            (*next)++;
+            break;
+
+        case t_id: {
+            // Could be function call or expression
+            Type id_type;
+            if (findIDType(curToken->data, &id_type)) {
+                switch (id_type) {
+                    case jf_void:
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/b_stmt.c:43]: no check for jf_string, what about concat or charat?

File: Parser/Tree_Nodes/b_stmt.c [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+            check = &g_array_index(token_stream, Token, (*next));
+            if (check->type == t_end_bracket) {
+                (*next)++;
+                new_statement->whileLoop = create_while_node(while_conditional, while_body);
+                break;
+            } else {
+                fprintf(stderr, "Syntax Error: Missing } at end of while statement");
+                exit(-1);
+            }
+
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/b_stmt.c]: All of this should be in while_node.c

File: Parser/Tree_Nodes/b_stmt.c [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+            b_stmt_list *for_body = create_b_stmt_list(token_stream, *next, next);
+            check = &g_array_index(token_stream, Token, (*next));
+            if (check->type == t_end_bracket) {
+                (*next)++;
+                new_statement->forLoop = create_for_node(for_initialize, for_conditional, for_incrementer, for_body);
+                break;
+            } else {
+                fprintf(stderr, "Syntax Error: Missing } at end of for statement");
+                exit(-1);
+            }
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/b_stmt.c]: All of this should be in for_node.c

File: Parser/Tree_Nodes/b_stmt.c [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+                        fprintf(stderr, "Syntax Error: Missing } at end of else statement");
+                        exit(-1);
+
+                    }
+                }
+                break;
+            } else {
+                fprintf(stderr, "Syntax Error: Missing } at end of if statement");
+                exit(-1);
+            }
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/b_stmt.c]: All of this should be in if_node.c

File: Parser/Tree_Nodes/i_expr.c [hunk 11]
```diff
@@ -174,7 +183,7 @@ i_expr * internal_i_expr_constructor(i_expr * parent, GArray * token_stream, uns
             curIndex++;
             curToken = &g_array_index(token_stream, Token, curIndex);
         }
-        if (curToken->type == t_end_paren || curToken->type == t_end_stmt || curToken->type == t_comma)
+        if (nextToken->type == t_end_paren || nextToken->type == t_end_stmt || nextToken->type == t_comma)
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/i_expr.c:199]: Why was this changed from curToken to nextToken?

File: Parser/Tree_Nodes/r_asmt.c [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    /**
+     * Token[index] is ID
+     * Token[index+1] is '='
+     * Token[index+2 <--> index+n] is expression, need to confirm result is same type
+     * Token[index+n+1] is ';'
+     */
+
+    r_asmt *new_asmt = calloc(1, sizeof(r_asmt));
+
+    // Need to try expression before we add the ID, because it cannot be self-referencing
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/r_asmt.c]: stale comment? We should not need to add any IDs when reassigning

File: Parser/Tree_Nodes/r_asmt.c [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    new_asmt->expression = create_expr(tokenStream, index + 2, next);
+
+    Type *varType = malloc(sizeof(Type));
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/r_asmt.c]: Unneeded and causes a memory leak, you can just create a `Type` variable and reference it in the function call, that way it will be removed from the stack frame when the function returns

File: Parser/Tree_Nodes/r_asmt.c [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Get the variable in the reassignment
+    Token *curToken = &g_array_index(tokenStream, Token, index);
+
+    // Does the ID exist, because we cant reassign a non existant assignment
+    if (!findIDType(curToken->data, varType)) {
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/r_asmt.c]: Need to use varType to confirm that the expression type matches the ID type

File: Parser/Tree_Nodes/stmt.c [hunk 15]
```diff
@@ -93,9 +101,16 @@ stmt * create_stmt(GArray * token_stream, unsigned long index, unsigned long * n
             // 2. Assignment
             new_statement->assignment = create_asmt(token_stream, index, next);
             break;
+
+            // Conditional/loops
+        case t_while:
+        case t_if:
+        case t_for:
+            new_statement->bStmtList = create_b_stmt_list(token_stream, index, next);
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/stmt.c]: These should call the individual constructors for each node, not a b_stmt_list.

File: Parser/Tree_Nodes/stmt.c [hunk 16]
```diff
@@ -63,15 +64,22 @@ stmt * create_stmt(GArray * token_stream, unsigned long index, unsigned long * n
...
                     case jstring:
                         // Expression
-                        new_statement->expression = create_expr(token_stream, index, next);
+                        curToken = &g_array_index(token_stream, Token, *next);
+                        Type t;
+                        if (findIDType(curToken->data, &t)) {
+                            new_statement->re_asmt = create_r_asmt(token_stream, index, next);
+                        } else {
+                            new_statement->expression = create_expr(token_stream, index, next);
+                        }
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/stmt.c]: Not sure what is going on here. If the first token is an ID then it MUST be a reassign. If the first token is a type ("Integer", "Double", "String") then it is an assignment. Do the checking for ID existaning in the reasssignment constructor.

File: Parser/Tree_Nodes/stmt.h [hunk 17]
```diff
@@ -28,8 +28,10 @@ typedef struct stmt
 {
     f_call * function_call;
     asmt * assignment;
+    r_asmt *re_asmt;
     expr * expression;
-}stmt;
+    b_stmt_list *bStmtList;
+} stmt;
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/stmt.h]: I don't think we described it in the meeting, but the stmt struct should not have a b_stmt_list, it should have elements for an if_node, for_node, and while_node, like b_stmt. The ONLY difference between the two is the lack of asmt in the b_stmt.

**Review by Reed-Terdal — CHANGES_REQUESTED** [review]

**Review by jxo7257 — DISMISSED** [review]

File: Parser/Tree_Nodes/for_node.c [hunk 18]
```diff
@@ -16,20 +16,72 @@
...
+
+    // Make new for node
     retVal = calloc(1, sizeof(for_node));
-    retVal->initialize = initalizer;
-    retVal->conditional = conditional;
-    retVal->incrementer = incrementer;
-    retVal->body = body;
-    return retVal;
+    (*next)++;
+    Token *check = &g_array_index(token_stream, Token, *next);
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/for_node.c]: You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

File: Parser/Tree_Nodes/if_node.c [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+#include <stdio.h>
+#include "if_node.h"
+#include "b_stmt_list.h"
+
+
+if_node *create_if_node(GArray *token_stream, unsigned long index, unsigned long *next) {
+    if_node *retVal = NULL;
+    retVal = calloc(1, sizeof(if_node));
+    (*next)++;
+    Token *check = &g_array_index(token_stream, Token, *next);
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/if_node.c]: You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

File: Parser/Tree_Nodes/r_asmt.c [hunk 20]
```diff
@@ -24,27 +24,45 @@ r_asmt *create_r_asmt(GArray *tokenStream, unsigned long index, unsigned long *n
...
-    // Need to try expression before we add the ID, because it cannot be self-referencing
-    new_asmt->expression = create_expr(tokenStream, index + 2, next);
-
-    Type *varType = malloc(sizeof(Type));
-
+    //Get expression so we know the type
     // Get the variable in the reassignment
-    Token *curToken = &g_array_index(tokenStream, Token, index);
-
+    Token *curToken = &g_array_index(tokenStream, Token, (*next));
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/r_asmt.c:29]: You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

File: Parser/Tree_Nodes/r_asmt.c [hunk 21]
```diff
@@ -24,27 +24,45 @@ r_asmt *create_r_asmt(GArray *tokenStream, unsigned long index, unsigned long *n
...
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
+    if (t == jstring || t == jf_str) {
+        if (new_asmt->expression->string_expression == NULL) {
+            fprintf(stderr, "Syntax Error: Incorrect reassignment, types do not match.");
+            exit(-1);
+        }
+    }
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/r_asmt.c:65]: This looks good, only thing is I would verify that the re-assignment ends with a semicolon, this is where you would use the `next` pointer.

File: Parser/Tree_Nodes/while_node.c [hunk 22]
```diff
@@ -23,15 +23,57 @@
...
 
     // Make new While node
     retVal = calloc(1, sizeof(while_node));
-    retVal->body = body;
-    retVal->conditional = conditional;
-    //Return the value
-    return retVal;
+    (*next)++;
+    // Load next token
+    Token *check = &g_array_index(token_stream, Token, *next);
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/while_node.c]: You should be using index, not next. Next is really only used after you create something with an unknown length. For example, when creating an assigmnment, we create the expression, then use next to make sure the following token is a semicolon

**Review by Reed-Terdal — CHANGES_REQUESTED** [review]
Need to becareful how you are using the next pointer, it should only be used after something of variable size.

**Review by jxo7257 — DISMISSED** [review]

File: Parser/Tree_Nodes/expr.c [hunk 23]
```diff
@@ -47,10 +46,22 @@ expr * create_expr(GArray * token_stream, unsigned long index, unsigned long * n
             break;
         default:
             // Unexpected Token when building expression
-            fprintf(stderr, "Syntax Error: Unexpected token when building expression");
+            fprintf(stderr, "Syntax Error: Unexpected token when building expression %s", first->data->str);
             exit(-1);
     }
 
+    Token *cur = &g_array_index(token_stream, Token, index + 1);
```
Comment (Reed-Terdal) [inline Parser/Tree_Nodes/expr.c:53]: Make sure to account for more complex expressions infront of comparison

**Review by Reed-Terdal — APPROVED** [review]

**Review by jxo7257 — APPROVED** [review]
